porch/pkg/registry/porch/secret.go

```diff
@@ -17,20 +17,41 @@ package porch
 import (
 	"context"
 	"fmt"
+	"sync"
+	"time"
 
+	"github.com/GoogleContainerTools/kpt/porch/pkg/registry/porch/wi"
 	"github.com/GoogleContainerTools/kpt/porch/pkg/repository"
+	"golang.org/x/oauth2"
+	stsv1 "google.golang.org/api/sts/v1"
 	core "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-func NewCredentialResolver(coreClient client.Reader) repository.CredentialResolver {
+const (
+	BasicAuthType            = core.SecretTypeBasicAuth
+	WorkloadIdentityAuthType = "kpt.dev/workload-identity-auth"
+	WIGCPSAAnnotation        = "iam.gke.io/gcp-service-account"
+)
+
+func NewCredentialResolver(coreClient client.Reader, resolverChain []Resolver) repository.CredentialResolver {
 	return &secretResolver{
-		coreClient: coreClient,
+		coreClient:    coreClient,
+		resolverChain: resolverChain,
 	}
 }
 
 type secretResolver struct {
-	coreClient client.Reader
+	resolverChain []Resolver
+	coreClient    client.Reader
+}
+
+type Resolver interface {
+	Resolve(ctx context.Context, secret core.Secret) (repository.Credential, bool, error)
```

Comment:
hmm...with this interface, secret sort of becomes essential. Another option could be to have a higher level struct say `authInfo` and authinfo is captured in the repo itself. So, in cases, where I am only using GCP WI, I can completely remove the need for secret and their RBAC etc.

```
type AuthInfo struct {
    type AuthType
    BasicAuth *BasicAuthInfo
    WIAuth *WIAuthInfo
}
```

And `secretResolver` itself has a mapping of `authType` --> porch.Resolver

Reply:
The auth structure is currently tightly linked to the secret. Agree we probably need to change this at some point, but I didn't see a need to do it now since all our current alternatives use the secret.